# Carregando arquivo de dados

In [2]:
with open('corretor-master/artigos.txt', 'r', errors = 'ignore', encoding = 'utf-8') as f:
    artigos = f.read()

In [3]:
print(artigos[:500])




imagem 

Temos a seguinte classe que representa um usuário no nosso sistema:

java

Para salvar um novo usuário, várias validações são feitas, como por exemplo: Ver se o nome só contém letras, [**o CPF só números**] e ver se o usuário possui no mínimo 18 anos. Veja o método que faz essa validação:

java 

Suponha agora que eu tenha outra classe, a classe `Produto`, que contém um atributo nome e eu quero fazer a mesma validação que fiz para o nome do usuário: Ver se só contém letras. E aí? Vou


In [4]:
len(artigos)

2605046

In [5]:
texto_exemplo = "Olá, tudo bem?"
tokens = texto_exemplo.split()

In [6]:
tokens

['Olá,', 'tudo', 'bem?']

In [7]:
len(tokens)

3

# NLTK separando os tokens

In [8]:
import nltk

In [9]:
#nltk.download('punkt')

In [10]:
# Separa as palavras dos caracteres
palavras_separadas = nltk.tokenize.word_tokenize(texto_exemplo)

In [11]:
palavras_separadas

['Olá', ',', 'tudo', 'bem', '?']

In [12]:
len(palavras_separadas)

5

In [13]:
'palavra'.isalpha()

True

In [14]:
# Percorre todos os dados do parâmetro, e atribui a uma lista apenas as palavras que são Alpha, que conteém letras
def separa_palavras(lista_tokens):
    lista_palavras = []
    for token in lista_tokens:
        if token.isalpha():
            lista_palavras.append(token)
    return lista_palavras

In [15]:
# Testando a função
separa_palavras(palavras_separadas)

['Olá', 'tudo', 'bem']

In [16]:
# Executando a função com os dados do 'artigo'
# variável que recebe as funcões de tokenização da base 'artigos'
lista_tokens = nltk.tokenize.word_tokenize(artigos)

# Executa a função com a variável que recebeu a tokenização
lista_palavras = separa_palavras(lista_tokens)

# Exibe a quantidade de palavras encontradas apenas com letras na base 'artigos'
print('O número de palavras é', len(lista_palavras))

O número de palavras é 403106


# Processo de normalização

### Retirando palavras repetidas

In [17]:
# função que recebe a 'lista_palavras', trasnforma a palavra para que fique toda minúscula 
def normalizacao(lista_palavras):
    lista_normalizada = []
    for palavra in lista_palavras:
        lista_normalizada.append(palavra.lower())
    return lista_normalizada

In [18]:
lista_normalizada = normalizacao(lista_palavras)
lista_normalizada[:500]

['imagem',
 'temos',
 'a',
 'seguinte',
 'classe',
 'que',
 'representa',
 'um',
 'usuário',
 'no',
 'nosso',
 'sistema',
 'java',
 'para',
 'salvar',
 'um',
 'novo',
 'usuário',
 'várias',
 'validações',
 'são',
 'feitas',
 'como',
 'por',
 'exemplo',
 'ver',
 'se',
 'o',
 'nome',
 'só',
 'contém',
 'letras',
 'o',
 'cpf',
 'só',
 'números',
 'e',
 'ver',
 'se',
 'o',
 'usuário',
 'possui',
 'no',
 'mínimo',
 'anos',
 'veja',
 'o',
 'método',
 'que',
 'faz',
 'essa',
 'validação',
 'java',
 'suponha',
 'agora',
 'que',
 'eu',
 'tenha',
 'outra',
 'classe',
 'a',
 'classe',
 'produto',
 'que',
 'contém',
 'um',
 'atributo',
 'nome',
 'e',
 'eu',
 'quero',
 'fazer',
 'a',
 'mesma',
 'validação',
 'que',
 'fiz',
 'para',
 'o',
 'nome',
 'do',
 'usuário',
 'ver',
 'se',
 'só',
 'contém',
 'letras',
 'e',
 'aí',
 'vou',
 'criar',
 'outro',
 'método',
 'para',
 'fazer',
 'a',
 'mesma',
 'validação',
 'ou',
 'criar',
 'uma',
 'interface',
 'ou',
 'uma',
 'classe',
 'que',
 'tanto',
 'usuario

### Texto normalizado, todas as palavras que contém apenas letras foram separadas em uma lista, e trasnformada em minúscula.

In [19]:
# A funcção set, retorna os valores da lista sem repetições
set([4, 4, 0, 1, 2, 3, 3, 3, 5, 6, 6, 8, 9, 7])

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}

In [20]:
len(set(lista_normalizada))

18465

### De um total de 403106 palavras, fora reduzido para 18465 palavras sem reperições e sem caracteres

In [21]:
''' Função definida para receber uma pavara e gerar um fatiamento no lado esquedo e no lado direito da palavra,
    para que a letra de teste seja alocada entre elas e passando de posição em posição '''

palavra_exemplo = "lgica"
def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra) + 1):
        fatias.append((palavra[:i], palavra[i:]))
    print(fatias)
    #palavras_geradas = insere_letras(fatias)
    #return palavras_geradas

In [22]:
''' Exemplo do que irá ocorrer com a função em cada palavra, de início, toda a palavra estará no canto direito
    e a letra será inserida na posição 1, então cada letra da palavra se deslocara para a esquerda e a letra
    de complementação se encaixara em cada espaço dessa transição'''
gerador_palavras(palavra_exemplo)

[('', 'lgica'), ('l', 'gica'), ('lg', 'ica'), ('lgi', 'ca'), ('lgic', 'a'), ('lgica', '')]


In [38]:
''' Função que insere uma nova letra, percorrendo todas as letras da variável destacada, e concatenando com o resultado
    da função 'gerador_de_palavras' '''
palavra_exemplo = "lgica"
def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E,  D in fatias:    
        for letra in letras:  
            novas_palavras.append(E + letra + D)
    return novas_palavras


''' Função que fatia as palavras'''
def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)):
        fatias.append((palavra[:i], palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    return palavras_geradas

palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['algica', 'blgica', 'clgica', 'dlgica', 'elgica', 'flgica', 'glgica', 'hlgica', 'ilgica', 'jlgica', 'klgica', 'llgica', 'mlgica', 'nlgica', 'olgica', 'plgica', 'qlgica', 'rlgica', 'slgica', 'tlgica', 'ulgica', 'vlgica', 'wlgica', 'xlgica', 'ylgica', 'zlgica', 'àlgica', 'álgica', 'âlgica', 'ãlgica', 'èlgica', 'élgica', 'êlgica', 'ìlgica', 'ílgica', 'îlgica', 'òlgica', 'ólgica', 'ôlgica', 'õlgica', 'ùlgica', 'úlgica', 'ûlgica', 'çlgica', 'lagica', 'lbgica', 'lcgica', 'ldgica', 'legica', 'lfgica', 'lggica', 'lhgica', 'ligica', 'ljgica', 'lkgica', 'llgica', 'lmgica', 'lngica', 'logica', 'lpgica', 'lqgica', 'lrgica', 'lsgica', 'ltgica', 'lugica', 'lvgica', 'lwgica', 'lxgica', 'lygica', 'lzgica', 'làgica', 'lágica', 'lâgica', 'lãgica', 'lègica', 'légica', 'lêgica', 'lìgica', 'lígica', 'lîgica', 'lògica', 'lógica', 'lôgica', 'lõgica', 'lùgica', 'lúgica', 'lûgica', 'lçgica', 'lgaica', 'lgbica', 'lgcica', 'lgdica', 'lgeica', 'lgfica', 'lggica', 'lghica', 'lgiica', 'lgjica', 'lgkica', 'lglica',

# Corretor

In [24]:
# Retorna a palavra que tem maior probabilidade de ser a correta em relaçao ao gerador de palavras e a base 'artigos'
def corretor(palavra):
    palavras_geradas = gerador_palavras(palavra)
    palavra_correta = max(palavras_geradas, key = probabilidade)
    return palavra_correta

In [25]:
# retorna as palavras com maiores frequencias em 'artigos'
frequencia = nltk.FreqDist(lista_normalizada)
total_palavras = len(lista_normalizada)
frequencia.most_common(10)

[('de', 15502),
 ('o', 14056),
 ('que', 12230),
 ('a', 11099),
 ('e', 10501),
 ('para', 7710),
 ('um', 6368),
 ('é', 5899),
 ('uma', 5220),
 ('do', 5124)]

In [26]:
# retorna a probabilidade a palavra indicada, em relação a quantidade de palavras em 'artigos'
def probabilidade(palavra_gerada):
    return frequencia[palavra_gerada]/total_palavras

probabilidade("lógica")

0.00023815075935361914

In [27]:
# Resultado esperado com as funções, retorna todas as possível palavras, e retorna como resultado a com maior probabilidade
corretor('lóica')

'lógica'

In [28]:
# Resultado esperado com as funções, retorna todas as possível palavras, e retorna como resultado a com maior probabilidade
corretor('hoe')

'hoje'

### Este corretor funciona apenas quando a palavra a ser corrigida está faltando uma letra e as outras estão corretas, não funciona caso a palavra tenha uma letra  a mais, ou se está faltando uma letra e no meio da palavra ainda existe outra letra errada como por exemplo a falta de um acento, e vale ressaltar que são probabilidades, o que não existe certeza absoluta da palavra retornada pelo corretor ser a esperada.

# Avaliador

In [29]:
def cria_dados_teste(nome_arquivo):
    lista_palavras_teste = []
    f = open(nome_arquivo, "r", errors = 'ignore', encoding = 'utf-8')
    for linha in f:
        correta, errada = linha.split()
        lista_palavras_teste.append((correta, errada))
    f.close()
    return lista_palavras_teste

lista_teste = cria_dados_teste('corretor-master/palavras.txt')

In [30]:
#print(lista_teste)

In [31]:
    
def avaliador(testes):
    numero_palavras = len(testes)
    acertou = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = round(acertou*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras")

In [32]:
avaliador(lista_teste)

1.08% de 186 palavras


# Deletando caracteres

In [33]:
# função para deletar um caractere e concatenar 
def deletando_caracteres(fatias):
    novas_palavras = []
    for E, D in fatias:
        novas_palavras.append(E + D[1:])
    return novas_palavras

In [35]:
def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)):
        fatias.append((palavra[:i], palavra[i:]))
    palavras_geradas = insere_letras(fatias) # insere o caractere
    palavras_geradas += deletando_caracteres(fatias) # deleta o caractere
    return palavras_geradas

palavra_exemplo = 'lógica'
palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['alógica', 'blógica', 'clógica', 'dlógica', 'elógica', 'flógica', 'glógica', 'hlógica', 'ilógica', 'jlógica', 'klógica', 'llógica', 'mlógica', 'nlógica', 'ológica', 'plógica', 'qlógica', 'rlógica', 'slógica', 'tlógica', 'ulógica', 'vlógica', 'wlógica', 'xlógica', 'ylógica', 'zlógica', 'àlógica', 'álógica', 'âlógica', 'ãlógica', 'èlógica', 'élógica', 'êlógica', 'ìlógica', 'ílógica', 'îlógica', 'òlógica', 'ólógica', 'ôlógica', 'õlógica', 'ùlógica', 'úlógica', 'ûlógica', 'çlógica', 'laógica', 'lbógica', 'lcógica', 'ldógica', 'leógica', 'lfógica', 'lgógica', 'lhógica', 'liógica', 'ljógica', 'lkógica', 'llógica', 'lmógica', 'lnógica', 'loógica', 'lpógica', 'lqógica', 'lrógica', 'lsógica', 'ltógica', 'luógica', 'lvógica', 'lwógica', 'lxógica', 'lyógica', 'lzógica', 'làógica', 'láógica', 'lâógica', 'lãógica', 'lèógica', 'léógica', 'lêógica', 'lìógica', 'líógica', 'lîógica', 'lòógica', 'lóógica', 'lôógica', 'lõógica', 'lùógica', 'lúógica', 'lûógica', 'lçógica', 'lóagica', 'lóbgica', 'lócgica'

In [36]:
avaliador(lista_teste)

41.4% de 186 palavras


### O algorítimo agora é capaz de além de adicionar, também deleta um caracter da palavra, com esses dois tratamentos conseguimos uma taxa de 41.4% de precisão até o momento.m 

# Trocando a letra

In [39]:
def troca_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E,  D in fatias:    
        for letra in letras:  
            novas_palavras.append(E + letra + D[1:])
    return novas_palavras

In [40]:
def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)):
        fatias.append((palavra[:i], palavra[i:]))
    palavras_geradas = insere_letras(fatias) # insere o caractere
    palavras_geradas += deletando_caracteres(fatias) # deleta o caractere
    palavras_geradas += troca_letras(fatias)
    return palavras_geradas

In [42]:
palavra_exemplo = 'lígica'
palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['alígica', 'blígica', 'clígica', 'dlígica', 'elígica', 'flígica', 'glígica', 'hlígica', 'ilígica', 'jlígica', 'klígica', 'llígica', 'mlígica', 'nlígica', 'olígica', 'plígica', 'qlígica', 'rlígica', 'slígica', 'tlígica', 'ulígica', 'vlígica', 'wlígica', 'xlígica', 'ylígica', 'zlígica', 'àlígica', 'álígica', 'âlígica', 'ãlígica', 'èlígica', 'élígica', 'êlígica', 'ìlígica', 'ílígica', 'îlígica', 'òlígica', 'ólígica', 'ôlígica', 'õlígica', 'ùlígica', 'úlígica', 'ûlígica', 'çlígica', 'laígica', 'lbígica', 'lcígica', 'ldígica', 'leígica', 'lfígica', 'lgígica', 'lhígica', 'liígica', 'ljígica', 'lkígica', 'llígica', 'lmígica', 'lnígica', 'loígica', 'lpígica', 'lqígica', 'lrígica', 'lsígica', 'ltígica', 'luígica', 'lvígica', 'lwígica', 'lxígica', 'lyígica', 'lzígica', 'làígica', 'láígica', 'lâígica', 'lãígica', 'lèígica', 'léígica', 'lêígica', 'lìígica', 'líígica', 'lîígica', 'lòígica', 'lóígica', 'lôígica', 'lõígica', 'lùígica', 'lúígica', 'lûígica', 'lçígica', 'líagica', 'líbgica', 'lícgica'

# Inverção de letras

In [43]:
def inverte_letras(fatias):
    novas_palavras = []
    for E, D in fatias:
        if len(D) > 1:
            novas_palavras.append(E + D[1] + D[0] + D [2:])
    return novas_palavras

In [44]:
palavra_exemplo = 'lgóica'
palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['algóica', 'blgóica', 'clgóica', 'dlgóica', 'elgóica', 'flgóica', 'glgóica', 'hlgóica', 'ilgóica', 'jlgóica', 'klgóica', 'llgóica', 'mlgóica', 'nlgóica', 'olgóica', 'plgóica', 'qlgóica', 'rlgóica', 'slgóica', 'tlgóica', 'ulgóica', 'vlgóica', 'wlgóica', 'xlgóica', 'ylgóica', 'zlgóica', 'àlgóica', 'álgóica', 'âlgóica', 'ãlgóica', 'èlgóica', 'élgóica', 'êlgóica', 'ìlgóica', 'ílgóica', 'îlgóica', 'òlgóica', 'ólgóica', 'ôlgóica', 'õlgóica', 'ùlgóica', 'úlgóica', 'ûlgóica', 'çlgóica', 'lagóica', 'lbgóica', 'lcgóica', 'ldgóica', 'legóica', 'lfgóica', 'lggóica', 'lhgóica', 'ligóica', 'ljgóica', 'lkgóica', 'llgóica', 'lmgóica', 'lngóica', 'logóica', 'lpgóica', 'lqgóica', 'lrgóica', 'lsgóica', 'ltgóica', 'lugóica', 'lvgóica', 'lwgóica', 'lxgóica', 'lygóica', 'lzgóica', 'làgóica', 'lágóica', 'lâgóica', 'lãgóica', 'lègóica', 'légóica', 'lêgóica', 'lìgóica', 'lígóica', 'lîgóica', 'lògóica', 'lógóica', 'lôgóica', 'lõgóica', 'lùgóica', 'lúgóica', 'lûgóica', 'lçgóica', 'lgaóica', 'lgbóica', 'lgcóica'

In [45]:
avaliador(lista_teste)

76.34% de 186 palavras
